src/coprocessor/dag/batch/executors/util/scan_executor.rs

```diff
@@ -236,13 +235,17 @@ impl<S: Store, I: ScanExecutorImpl, P: PointRangePolicy> BatchExecutor for ScanE
         }
     }
 
-    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
-        for (index, num) in self.scanned_rows_per_range.iter_mut().enumerate() {
-            destination.scanned_rows_per_range[index] += *num;
-            *num = 0;
-        }
+    #[inline]
+    fn collect_exec_stats(&mut self, dest: &mut ExecuteStats) {
+        dest.scanned_rows_per_range
+            .append(&mut self.scanned_rows_per_range);
+        self.scanned_rows_per_range.push(0);
```

Comment:
I have no idea why we need to append an extra `0` to the vector according to TiDB https://github.com/pingcap/tidb/blob/ebfb2ae7fef88447d265e1df5cc7a298bbc307e5/distsql/stream.go#L91 and the inner logic of `QueryFeedback.Update`

Reply:
It is to prevent the following scenario:

1. receive new range
2. next() several times
3. stop scan (while the range is not drained)
4. collect exec stats and generate a stream request
5. start scan
6. next()

In step 6, it needs to perform `last count += 1` operation. Thus, appending a `0` in step 4 after previous counts are taken is a cheap solution. It is slightly faster than:

1. if empty: append 1
2. if not empty: last count += 1

Reply:
BTW TiDB does not actually rely on this "scanned rows each range".. It is only used for slow query output.

Reply:
BTW I was told that this count is only used in slow output.. (although according to your output looks like this is not true)..

Reply:
It's OK because the final result will be correct (even though some extra memory waste).

Reply:
There could be some edge case that might be wrong, as seen in the modified test. Consider:

1. receive new range
2. next() several times
3. stop scan
4. collect exec stats and generate a stream request
5. start scan
6. next() found that last range is drained so that a new range is entered and start iterating

After that, when we collect exec stats we will get `[0, x, ...]` instead of `[x, ...]`. I'm not sure whether or not this is compatible with TiDB's code, but it's current behaviour so I kept the same.